rmc/src/service.rs

```diff
@@ -0,0 +1,351 @@
+    RmcMessage,
+};
+pub use aleph_bft_crypto::{
+    Indexed, MultiKeychain, Multisigned, NodeCount, PartialMultisignature, PartiallyMultisigned,
+    Signable, Signature, Signed, UncheckedSigned,
+};
+use core::fmt::Debug;
+use log::{debug, warn};
+use std::hash::Hash;
+
+const LOG_TARGET: &str = "AlephBFT-rmc";
+
+/// Reliable Multicast Box
+///
+/// The instance of [`Service<H, MK, SCH>`] is used to reliably broadcasts hashes of type `H`.
+/// It collects the signed hashes and upon receiving a large enough number of them it yields
+/// the multisigned hash.
+///
+/// A node with an instance of [`Service<H, MK, SCH>`] can initiate broadcasting a message `msg: H`
+/// by calling [`Service::start_rmc`]. As a result, the node signs `msg` and starts scheduling
+/// messages for broadcast which can be obtained by awaiting on [`Service::next_message`]. When
+/// sufficiently many nodes initiate rmc with the same message `msg` and a node collects enough
+/// signatures to form a complete multisignature under the message, [`Service::process_message`]
+/// will return the multisigned hash.
+///
+/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html
+/// for a high-level description of this protocol and how it is used for fork alerts.
+pub struct Service<H, MK, SCH>
+where
+    H: Signable + Hash,
+    MK: MultiKeychain,
+    SCH: TaskScheduler<RmcMessage<H, MK::Signature, MK::PartialMultisignature>>,
+{
+    scheduler: SCH,
+    handler: Handler<H, MK>,
+}
+
+impl<H, MK, SCH> Service<H, MK, SCH>
+where
+    H: Signable + Hash + Eq + Clone + Debug,
+    MK: MultiKeychain,
+    SCH: TaskScheduler<RmcMessage<H, MK::Signature, MK::PartialMultisignature>>,
+{
+    pub fn new(scheduler: SCH, handler: Handler<H, MK>) -> Self {
+        Service { scheduler, handler }
+    }
+
+    /// Signs the given `hash` and adds the signature to the collection. If the given `hash`
+    /// completes the multisignature, it is scheduled for the broadcasts and then returned.
+    /// If the multisignature is not completed, `None` is returned. If the multisignature was
+    /// already completed when starting rmc, no tasks are scheduled. Otherwise the signed hash
+    /// is scheduled for the broadcasts.
+    pub fn start_rmc(&mut self, hash: H) -> Option<Multisigned<H, MK>> {
+        debug!(target: LOG_TARGET, "starting rmc for {:?}", hash);
+        match self.handler.on_start_rmc(hash) {
+            OnStartRmcResponse::SignedHash(signed_hash) => {
+                self.scheduler
+                    .add_task(RmcMessage::SignedHash(signed_hash.into_unchecked()));
+            }
+            OnStartRmcResponse::MultisignedHash(multisigned) => {
+                self.scheduler.add_task(RmcMessage::MultisignedHash(
+                    multisigned.clone().into_unchecked(),
+                ));
+                return Some(multisigned);
+            }
+            OnStartRmcResponse::Noop => {}
+        }
+        None
```

Comment:
Hm, I think this might be marginally more readable if you have two of these at the end of the select branches, but don't have the `return` there. Then it's explicit what every branch returns, rather than having to parse the entire thing to understand that.

Reply:
uhhh, lemme change the other one as well